### PR TITLE
MRG: TD7

### PR DESCRIPTION
**Pull request by AlexanderFabisch**
TODO

- [x] details
    - [x] separate exploration noise and target policy noise (also in TD3?)
    - [x] value clipping
- [x] TD7 extension: checkpoints
- [x] check checkpointing logic and document it better
- [x] refactoring
    - [x] extract sale
    - [x] extract checkpointing
    - [x] create new nnx.module that makes sampling functions from ddpg/td3 reusable
- [x] documentation
- [x] add to api docs
- [x] unit test
- [x] evaluation
    - [x] check results
        - 4 runs, better than TD3 after 1M steps: ca. 3800 return on Hopper-v5
    - [x] rerun after correcting actor loss with 5M timesteps

## Results

| | [TD7 Paper](https://arxiv.org/pdf/2306.02451) | This PR |
| - | - | - |
| Training runs | 10 | 5 |
| Test episodes | 10 | 10 |
| Hopper - Return at 5M steps (+- $\frac{1.96}{\sqrt{N}} \sigma$) | 4075 +- 225 | 4164 +- 48 |

Learning curves:

Screenshot from 2025-06-24 17-54-30

### CODE REVIEW

**Comment by AlexanderFabisch** [comment]
@mlaux1 almost done, only waiting for the last 2 training runs, should take about a day to finish.

File: rl_blox/blox/embedding/sale.py [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+    -------
+    loss : float
+        Loss value.
+    """
+    zsa, _ = embedding(observation, action)
+    zsp = jax.lax.stop_gradient(embedding.state_embedding(next_observation))
+    return optax.squared_error(predictions=zsa, targets=zsp).mean()
+
+
+class DeterministicSALEPolicy(nnx.Module):
```
Comment (AlexanderFabisch) [inline rl_blox/blox/embedding/sale.py:236]: move to the top (after actor)

File: rl_blox/algorithm/td7.py [hunk 2]
```diff
@@ -0,0 +1,884 @@
...
+            "actor",
+            "actor_target",
+            "actor_optimizer",
+            "critic",
+            "critic_target",
+            "critic_optimizer",
+            "replay_buffer",
+        ],
+    )(
+        fixed_embedding_checkpoint if use_checkpoints else embedding,
```
Comment (AlexanderFabisch) [inline rl_blox/algorithm/td7.py:875]: Should still return fixed embedding when not checkpointing.

File: rl_blox/algorithm/td7.py [hunk 3]
```diff
@@ -0,0 +1,884 @@
...
+        embedding_optimizer,
```
Comment (AlexanderFabisch) [inline rl_blox/algorithm/td7.py:876]: Return embedding too!

File: rl_blox/algorithm/td7.py [hunk 4]
```diff
@@ -0,0 +1,884 @@
...
+        actor_checkpoint if use_checkpoints else actor,
```
Comment (AlexanderFabisch) [inline rl_blox/algorithm/td7.py:877]: And embedding_target!

File: rl_blox/algorithm/td7.py [hunk 5]
```diff
@@ -0,0 +1,884 @@
...
+    steps_before_checkpointing: int = 750_000,
+    reset_weight: float = 0.9,
+    batch_size: int = 256,
+    learning_starts: int = 25_000,
+    replay_buffer: LAP | None = None,
+    actor_target: ActorSALE | None = None,
+    critic_target: ContinuousClippedDoubleQNet | None = None,
+    logger: LoggerBase | None = None,
+) -> tuple[
+    nnx.Module,
```
Comment (AlexanderFabisch) [inline rl_blox/algorithm/td7.py:436]: Update signature, see below

**Comment by mlaux1** [comment]
I'll review and merge this tomorrow

**Review by mlaux1 — APPROVED** [review]
